### PR TITLE
Webauthn

### PR DESCRIPTION
**Pull request by civts**
PR to finalize the changes made in #51.

### CODE REVIEW

File: src/main/java/it/smartcommunitylab/aac/webauthn/service/WebAuthnRpService.java [hunk 1]
```diff
@@ -0,0 +1,269 @@
...
+                webAuthnUserAccountService.addAccount(account);
+
+                // Create credential in the database
+                WebAuthnCredential newCred = new WebAuthnCredential();
+                newCred.setCreatedOn(new Date());
+                newCred.setLastUsedOn(new Date());
+                newCred.setCredentialId(result.getKeyId().getId().getBase64());
+                newCred.setPublicKeyCose(result.getPublicKeyCose().getBase64());
+                newCred.setSignatureCount(result.getSignatureCount());
+                Optional<SortedSet<AuthenticatorTransport>> transportsOpt = result.getKeyId().getTransports();
```
Comment (matteo-s) [inline src/main/java/it/smartcommunitylab/aac/webauthn/service/WebAuthnRpService.java:158]: rewrite the block:
- we don't need the set as sorted
- use streams to map values
- let the field null if not set

```
          if (result.getKeyId().getTransports().isPresent()) {
                    Set<AuthenticatorTransport> transports = result.getKeyId().getTransports().get();
                    List<String> transportCodes = transports.stream().map(t -> t.getId()).collect(Collectors.toList());
                    newCred.setTransports(StringUtils.collectionToCommaDelimitedString(transportCodes));
                }
```

File: src/main/java/it/smartcommunitylab/aac/webauthn/service/WebAuthnRpService.java [hunk 2]
```diff
@@ -0,0 +1,269 @@
...
+                } else {
+                    newCred.setTransports("");
+                }
+                newCred.setUserHandle(account.getUserHandle());
+
+                webAuthnUserAccountService.saveCredential(newCred);
+                activeRegistrations.remove(key);
+                return username;
+            } else {
+                return null;
```
Comment (matteo-s) [inline src/main/java/it/smartcommunitylab/aac/webauthn/service/WebAuthnRpService.java:175]: remove else block, see above

File: src/main/java/it/smartcommunitylab/aac/webauthn/service/WebAuthnRpService.java [hunk 3]
```diff
@@ -0,0 +1,269 @@
...
+                    username);
+            if (existingAccount != null) {
+                throw new WebAuthnAuthenticationException("_",
+                        "Account already registered");
+            }
+            final PublicKeyCredentialCreationOptions options = info.getOptions();
+            RegistrationResult result = rp
+                    .finishRegistration(FinishRegistrationOptions.builder().request(options).response(pkc).build());
+            boolean attestationIsTrusted = result.isAttestationTrusted();
+            if (attestationIsTrusted) {
```
Comment (matteo-s) [inline src/main/java/it/smartcommunitylab/aac/webauthn/service/WebAuthnRpService.java:139]: if attestation is not trusted we should throw an exception

File: src/main/java/it/smartcommunitylab/aac/webauthn/service/WebAuthnRpService.java [hunk 4]
```diff
@@ -0,0 +1,269 @@
...
+     *         null
+     */
+    public String finishLogin(
+            PublicKeyCredential<AuthenticatorAssertionResponse, ClientAssertionExtensionOutputs> pkc,
+            String sessionId) throws WebAuthnAuthenticationException {
+        try {
+            AssertionRequest assertionRequest = activeAuthentications.get(sessionId);
+            AssertionResult result = rp.finishAssertion(FinishAssertionOptions.builder().request(assertionRequest)
+                    .response(pkc).build());
+            if (result.isSuccess() && result.isSignatureCounterValid()) {
```
Comment (matteo-s) [inline src/main/java/it/smartcommunitylab/aac/webauthn/service/WebAuthnRpService.java:217]: as above, if not success throw an exception

File: src/main/java/it/smartcommunitylab/aac/webauthn/service/WebAuthnYubicoCredentialsRepository.java [hunk 5]
```diff
@@ -0,0 +1,121 @@
...
+ * For this class, everytime we should use the 'username', we
+ * will instead use a string in the following format:
+ * 'realmname{@link #separator}username'. This
+ * is due to the fact that yubico's library assumes that two distinct users can
+ * not have the same username (https://git.io/JD5Vr).
+ */
+public class WebAuthnYubicoCredentialsRepository implements CredentialRepository {
+
+    private final String providerId;
+    private WebAuthnUserAccountService webAuthnUserAccountService;
```
Comment (matteo-s) [inline src/main/java/it/smartcommunitylab/aac/webauthn/service/WebAuthnYubicoCredentialsRepository.java:32]: field is final

**Review by matteo-s — CHANGES_REQUESTED** [review]

File: pom.xml [hunk 6]
```diff
@@ -281,6 +281,12 @@
 			<artifactId>spring-boot-configuration-processor</artifactId>
 			<optional>true</optional>
 		</dependency>
+		<dependency>
+			<groupId>com.yubico</groupId>
+			<artifactId>webauthn-server-core</artifactId>
+			<version>1.12.1</version>
+			<scope>compile</scope>
```
[the remark this answers is not in the document]
Comment (civts) [reply pom.xml]: Fixed in bac191a1ade8aff623818557bf938856c2b941f1

File: src/main/java/it/smartcommunitylab/aac/config/AACConfig.java [hunk 7]
```diff
@@ -402,4 +408,4 @@ public RealmAwarePathUriBuilder realmUriBuilder() {
 //        return new InternalAuthorityHandler();
 //    }
 
-}
+}
```
[the remark this answers is not in the document]
Comment (civts) [reply src/main/java/it/smartcommunitylab/aac/config/AACConfig.java]: Done with 5b2ed1453f3216fc0aa129518dc973f3171f864a

File: src/main/java/it/smartcommunitylab/aac/config/DatabaseConfig.java [hunk 8]
```diff
@@ -69,7 +68,8 @@
         "it.smartcommunitylab.aac.services.persistence",
         "it.smartcommunitylab.aac.attributes.persistence",
         "it.smartcommunitylab.aac.repository",
-}, queryLookupStrategy = QueryLookupStrategy.Key.CREATE_IF_NOT_FOUND)
+        "it.smartcommunitylab.aac.webauthn.persistence"},
```
[the remark this answers is not in the document]
Comment (civts) [reply src/main/java/it/smartcommunitylab/aac/config/DatabaseConfig.java]: Done with ed7a292970e7584a25bd8bd3c71739d58173e16a

File: src/main/java/it/smartcommunitylab/aac/core/AuthorityManager.java [hunk 9]
```diff
@@ -4,8 +4,10 @@
 import java.util.Collection;
 import java.util.Collections;
 import java.util.List;
+
```
[the remark this answers is not in the document]
Comment (civts) [reply src/main/java/it/smartcommunitylab/aac/core/AuthorityManager.java]: Done with 0d77d99dc1c675f3d961fced20ae1531de081521

File: src/main/java/it/smartcommunitylab/aac/webauthn/auth/WebAuthnAuthenticationException.java [hunk 10]
```diff
@@ -0,0 +1,79 @@
...
+import it.smartcommunitylab.aac.SystemKeys;
+
+public class WebAuthnAuthenticationException extends AuthenticationException {
+ 
+    private static final long serialVersionUID = SystemKeys.AAC_CORE_SERIAL_VERSION;
+
+    private final String subject;
+    private final String username;
+    private final String credentials;
+    private final String flow;
```
[the remark this answers is not in the document]
Comment (civts) [reply src/main/java/it/smartcommunitylab/aac/webauthn/auth/WebAuthnAuthenticationException.java]: Definitely a leftover from the original class, `InternalAuthenticationException`. Removed it in 6680ba971b603e39d3d6475f7929e031d4d3ecc5

File: src/main/java/it/smartcommunitylab/aac/webauthn/controller/WebAuthnAuthenticationController.java [hunk 11]
```diff
@@ -0,0 +1,115 @@
...
+    @PostMapping(value = "/auth/webauthn/assertionOptions/{providerId}", consumes = MediaType.APPLICATION_JSON_VALUE, produces = MediaType.APPLICATION_JSON_VALUE)
+    @ResponseBody
+    public WebAuthnLoginResponse generateAssertionOptions(@RequestBody @Valid WebAuthnAuthenticationStartRequest body,
+            @PathVariable("providerId") String providerId) {
+        try {
+            WebAuthnRpService rps = webAuthnRpServiceRegistrationRepository.get(providerId);
+
+            String username = body.getUsername();
+            if (!isValidUsername(username)) {
+                throw new ResponseStatusException(HttpStatus.BAD_REQUEST, "Invalid username");
```
[the remark this answers is not in the document]
Comment (civts) [reply src/main/java/it/smartcommunitylab/aac/webauthn/controller/WebAuthnAuthenticationController.java]: Done in d83c5fa36550b058adc8110a95254d740d193359

File: src/main/java/it/smartcommunitylab/aac/webauthn/controller/WebAuthnRegistrationController.java [hunk 12]
```diff
@@ -0,0 +1,182 @@
...
+     * Ensure the request mapping value matches the one returned from
+     * {@link WebAuthnIdentityService#getRegistrationUrl}
+     */
+    @Hidden
+    @RequestMapping(value = "/auth/webauthn/register/{providerId}", method = RequestMethod.GET)
+    public String registrationPage(@PathVariable("providerId") String providerId) {
+        try {
+            WebAuthnIdentityProviderConfig provider = registrationRepository.findByProviderId(providerId);
+            boolean canRegister;
+            if (provider != null) {
```
[the remark this answers is not in the document]
Comment (civts) [reply src/main/java/it/smartcommunitylab/aac/webauthn/controller/WebAuthnRegistrationController.java]: Done in d83c5fa36550b058adc8110a95254d740d193359 (added a more relevant message in 7519372c932f8092bace8d0c160a9f9d16fc61f5)

File: src/main/java/it/smartcommunitylab/aac/webauthn/controller/WebAuthnRegistrationController.java [hunk 13]
```diff
@@ -0,0 +1,182 @@
...
+                canRegister = provider
+                        .getConfigMap()
+                        .isEnableRegistration();
+            } else {
+                canRegister = false;
+            }
+            if (!canRegister) {
```
[the remark this answers is not in the document]
Comment (civts) [reply src/main/java/it/smartcommunitylab/aac/webauthn/controller/WebAuthnRegistrationController.java]: Done in d83c5fa36550b058adc8110a95254d740d193359

File: src/main/java/it/smartcommunitylab/aac/webauthn/controller/WebAuthnRegistrationController.java [hunk 14]
```diff
@@ -0,0 +1,182 @@
...
+                throw new RegistrationException("registration is disabled");
+            }
+            return "webauthn/register";
+        } catch (RegistrationException e) {
```
[the remark this answers is not in the document]
Comment (civts) [reply src/main/java/it/smartcommunitylab/aac/webauthn/controller/WebAuthnRegistrationController.java]: Done in d83c5fa36550b058adc8110a95254d740d193359

File: src/main/java/it/smartcommunitylab/aac/webauthn/controller/WebAuthnRegistrationController.java [hunk 15]
```diff
@@ -0,0 +1,182 @@
...
+            if (!isValidUsername(username)) {
+                throw new ResponseStatusException(HttpStatus.BAD_REQUEST, "Invalid username");
+            }
+
+            String displayName = body.getDisplayName();
+            if (!isValidDisplayName(displayName)) {
+                throw new ResponseStatusException(HttpStatus.BAD_REQUEST, "Invalid displayName");
+            }
+
+            WebAuthnIdentityProviderConfig providerCfg = registrationRepository.findByProviderId(providerId);
```
[the remark this answers is not in the document]
Comment (civts) [reply src/main/java/it/smartcommunitylab/aac/webauthn/controller/WebAuthnRegistrationController.java]: Addressed in 7bfb5c975c262a8843e3d90e9713587276e81a6f

File: src/main/java/it/smartcommunitylab/aac/webauthn/controller/WebAuthnRegistrationController.java [hunk 16]
```diff
@@ -0,0 +1,182 @@
...
+            String realm = providerCfg.getRealm();
+            final WebAuthnRegistrationResponse response = rps.startRegistration(
+                    username,
+                    realm,
```
[the remark this answers is not in the document]
Comment (civts) [reply src/main/java/it/smartcommunitylab/aac/webauthn/controller/WebAuthnRegistrationController.java]: Addressed in 7bfb5c975c262a8843e3d90e9713587276e81a6f

File: src/main/java/it/smartcommunitylab/aac/webauthn/controller/WebAuthnRegistrationController.java [hunk 17]
```diff
@@ -0,0 +1,182 @@
...
+    /**
+     * Validates the attestation generated using the Credential Creation
+     * Options obtained through the {@link #generateAttestationOptions} controller
+     */
+    @Hidden
+    @PostMapping(value = "/auth/webauthn/attestations/{providerId}", consumes = MediaType.APPLICATION_JSON_VALUE)
+    @ResponseBody
+    public String verifyAttestation(@RequestBody @Valid WebAuthnAttestationResponse body,
+            @PathVariable("providerId") String providerId) {
+        final ResponseStatusException invalidAttestationException = new ResponseStatusException(HttpStatus.BAD_REQUEST,
```
[the remark this answers is not in the document]
Comment (civts) [reply src/main/java/it/smartcommunitylab/aac/webauthn/controller/WebAuthnRegistrationController.java]: ok, done in 792e908e6a66126d76b73c07b1ceec7300d6c679

File: src/main/java/it/smartcommunitylab/aac/webauthn/model/WebAuthnAssertionResponse.java [hunk 18]
```diff
@@ -0,0 +1,50 @@
...
+    public Map<String, Serializable> getAssertion() {
+        return assertion;
+    }
+
+    public void setAssertion(Map<String, Serializable> assertion) {
+        this.assertion = assertion;
+    }
+
+    @JsonIgnore
+    public String getAssertionAsJson() throws JsonProcessingException{
```
[the remark this answers is not in the document]
Comment (civts) [reply src/main/java/it/smartcommunitylab/aac/webauthn/model/WebAuthnAssertionResponse.java]: Done in 86915d382eee27c9856bccd4d7d5d5f1c89c320a

File: src/main/java/it/smartcommunitylab/aac/webauthn/model/WebAuthnAttestationResponse.java [hunk 19]
```diff
@@ -0,0 +1,47 @@
...
+
+    public Map<String, Serializable> getAttestation() {
+        return attestation;
+    }
+
+    public void setAttestation(Map<String, Serializable> attestation) {
+        this.attestation = attestation;
+    }
+
+    public String exportAttestation() throws JsonProcessingException{
```
[the remark this answers is not in the document]
Comment (civts) [reply src/main/java/it/smartcommunitylab/aac/webauthn/model/WebAuthnAttestationResponse.java]: Done in 9c8d1d0d29be4c103b26cca3cec89a074ebc6c63

File: src/main/java/it/smartcommunitylab/aac/webauthn/model/UserWebAuthnCredentials.java [hunk 20]
```diff
@@ -0,0 +1,60 @@
...
+
+import javax.validation.Valid;
+import javax.validation.constraints.NotBlank;
+
+import com.fasterxml.jackson.annotation.JsonIgnore;
+
+import it.smartcommunitylab.aac.core.model.UserCredentials;
+
+@Valid
+public class UserWebAuthnCredentials implements UserCredentials {
```
[the remark this answers is not in the document]
Comment (civts) [reply src/main/java/it/smartcommunitylab/aac/webauthn/model/UserWebAuthnCredentials.java]: Done, 78b2361f3d41f0de2f7b8a262454ef424cb330f2

File: src/main/java/it/smartcommunitylab/aac/webauthn/model/WebAuthnLoginResponse.java [hunk 21]
```diff
@@ -0,0 +1,47 @@
...
+@Valid
+@JsonIgnoreProperties(ignoreUnknown = true)
+public class WebAuthnLoginResponse {
+
+    @JsonProperty("key")
+    @NotNull
+    private String key;
+
+    @NotNull
+    private AssertionRequest assertionrequest;
```
[the remark this answers is not in the document]
Comment (civts) [reply src/main/java/it/smartcommunitylab/aac/webauthn/model/WebAuthnLoginResponse.java]: Fixed in 565465c7749a2a5feededa9f96256759cdaaf1d1

File: src/main/java/it/smartcommunitylab/aac/webauthn/model/WebAuthnLoginResponse.java [hunk 22]
```diff
@@ -0,0 +1,47 @@
...
+
+    private static final ObjectMapper mapper = new ObjectMapper();
```
[the remark this answers is not in the document]
Comment (civts) [reply src/main/java/it/smartcommunitylab/aac/webauthn/model/WebAuthnLoginResponse.java]: Done, d91419230eccd52677dee76b3dc7ff558ba7443c

File: src/main/java/it/smartcommunitylab/aac/webauthn/model/WebAuthnRegistrationResponse.java [hunk 23]
```diff
@@ -0,0 +1,48 @@
...
+
+@Valid
+@JsonIgnoreProperties(ignoreUnknown = true)
+public class WebAuthnRegistrationResponse {
+
+    @JsonProperty("key")
+    @NotNull
+    private String key;
+
+    private static final ObjectMapper mapper = new ObjectMapper();
```
[the remark this answers is not in the document]
Comment (civts) [reply src/main/java/it/smartcommunitylab/aac/webauthn/model/WebAuthnRegistrationResponse.java]: Done, d91419230eccd52677dee76b3dc7ff558ba7443c

File: src/main/java/it/smartcommunitylab/aac/webauthn/service/WebAuthnRpService.java [hunk 24]
```diff
@@ -0,0 +1,269 @@
...
+
+    private final WebAuthnUserAccountService webAuthnUserAccountService;
+
+    private final SubjectService subjectService;
+
+    private final RelyingParty rp;
+    private final String provider;
+
+    private static final Long TIMEOUT = 9000L;
+    private final StringKeyGenerator keyGenerator = new Base64StringKeyGenerator(Base64.getUrlEncoder(), 64);
```
[the remark this answers is not in the document]
Comment (civts) [reply src/main/java/it/smartcommunitylab/aac/webauthn/service/WebAuthnRpService.java]: Done in 68b34de7ee6ef4b18754c495c03db706bd8a7c86

File: src/main/java/it/smartcommunitylab/aac/webauthn/service/WebAuthnRpService.java [hunk 25]
```diff
@@ -0,0 +1,269 @@
...
+            String realm,
+            String key) throws WebAuthnAuthenticationException {
+        try {
+            final WebAuthnCredentialCreationInfo info = activeRegistrations.get(key);
+            WebAuthnAuthenticationException noSuchRegistration = new WebAuthnAuthenticationException("_",
+                    "Can not find matching active registration request");
+            if (info == null) {
+                throw noSuchRegistration;
+            }
+            if (!info.getRealm().equals(realm)) {
```
[the remark this answers is not in the document]
Comment (civts) [reply src/main/java/it/smartcommunitylab/aac/webauthn/service/WebAuthnRpService.java]: Done in 7bfb5c975c262a8843e3d90e9713587276e81a6f

File: src/main/java/it/smartcommunitylab/aac/webauthn/service/WebAuthnRpService.java [hunk 26]
```diff
@@ -0,0 +1,269 @@
...
+            // In that case, we should allow registering multiple credentials
+            throw new WebAuthnAuthenticationException("_", "User already exists");
+        }
+        final UserIdentity user = generateUserIdentity(username, displayName);
+        final StartRegistrationOptions startRegistrationOptions = StartRegistrationOptions.builder().user(user)
+                .authenticatorSelection(authenticatorSelection).timeout(TIMEOUT).build();
+        final PublicKeyCredentialCreationOptions options = rp.startRegistration(startRegistrationOptions);
+        final WebAuthnCredentialCreationInfo info = new WebAuthnCredentialCreationInfo();
+        info.setUsername(username);
+        info.setRealm(realm);
```
[the remark this answers is not in the document]
Comment (civts) [reply src/main/java/it/smartcommunitylab/aac/webauthn/service/WebAuthnRpService.java]: Done in 7bfb5c975c262a8843e3d90e9713587276e81a6f

File: src/main/java/it/smartcommunitylab/aac/webauthn/service/WebAuthnRpService.java (hunk 25 above)
[the remark this answers is not in the document]
Comment (civts) [reply src/main/java/it/smartcommunitylab/aac/webauthn/service/WebAuthnRpService.java]: Done with dca1d21a3bca80ecab6cf65006a75da366d92c0b

File: src/main/java/it/smartcommunitylab/aac/webauthn/service/WebAuthnRpService.java [hunk 27]
```diff
@@ -0,0 +1,269 @@
...
+
+    /**
+     * Returns:
+     * - the username of the authenticated user on successful authentication
+     * - null if the authentication was not successful
+     * - throws a WebAuthnAuthenticationException if some other error occourred
+     */
+    public String finishRegistration(
+            PublicKeyCredential<AuthenticatorAttestationResponse, ClientRegistrationExtensionOutputs> pkc,
+            String realm,
```
[the remark this answers is not in the document]
Comment (civts) [reply src/main/java/it/smartcommunitylab/aac/webauthn/service/WebAuthnRpService.java]: Done in dca1d21a3bca80ecab6cf65006a75da366d92c0b (and removed realm in e13c590370a6ae0a264fd33e66e6934990fdb386)

File: src/main/java/it/smartcommunitylab/aac/webauthn/service/WebAuthnRpService.java (hunk 1 above)
Comment (civts) [reply src/main/java/it/smartcommunitylab/aac/webauthn/service/WebAuthnRpService.java]: Done, 744314d5e8dc67386a74144499be4f65c7f0ebad

File: src/main/java/it/smartcommunitylab/aac/webauthn/service/WebAuthnRpService.java (hunk 2 above)
Comment (civts) [reply src/main/java/it/smartcommunitylab/aac/webauthn/service/WebAuthnRpService.java]: Done in previous commits (I believe it was cb95a78885936ee33c8f8d0d9ad6d34aed10c7bf)

File: src/main/java/it/smartcommunitylab/aac/webauthn/service/WebAuthnRpService.java (hunk 3 above)
Comment (civts) [reply src/main/java/it/smartcommunitylab/aac/webauthn/service/WebAuthnRpService.java]: Addressed in cb95a78885936ee33c8f8d0d9ad6d34aed10c7bf

File: src/main/java/it/smartcommunitylab/aac/webauthn/service/WebAuthnYubicoCredentialsRepository.java (hunk 5 above)
Comment (civts) [reply src/main/java/it/smartcommunitylab/aac/webauthn/service/WebAuthnYubicoCredentialsRepository.java]: Addressed in 7083e6860230f9dfaab6f71b5ba7338caab4e077

File: src/main/java/it/smartcommunitylab/aac/webauthn/service/WebAuthnYubicoCredentialsRepository.java [hunk 28]
```diff
@@ -0,0 +1,121 @@
...
+
+    public WebAuthnYubicoCredentialsRepository(String provider,
+            WebAuthnUserAccountService webAuthnUserAccountService) {
+        this.webAuthnUserAccountService = webAuthnUserAccountService;
```
[the remark this answers is not in the document]
Comment (civts) [reply src/main/java/it/smartcommunitylab/aac/webauthn/service/WebAuthnYubicoCredentialsRepository.java:38]: Addressed in 7083e6860230f9dfaab6f71b5ba7338caab4e077

**Pull request merged**